### ISSUE TITLE
something about cross_iou_loss

### ISSUE THREAD

**Issue by Mr2er0** [Duankaiwen/LSNet#22]
[cross_iou_loss](https://github.com/Duankaiwen/LSNet/blob/7ca365cefdbab3a039fff160bc61acd0d16bb70c/code/mmdet/models/losses/cross_iou_loss.py#L62)
这里所对应的cross_iou_loss是文章的所提出的损失吗？有人可以告诉我这里面loss的输入的tensor的维度和格式大概都是什么吗？谢谢。

**Comment by zlj63501** [comment]
看了源码部分，很奇怪，感觉和论文的损失很不一样。
segm 的输入：pred    (2(batch)*h*w, 148)    148 指的是(36+1)*4
                        target  (2(batch)*h*w, 74*2)
期待作者能回复一下，感谢~

**Comment by Duankaiwen** [comment]
@Zlj-github 
36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742

**Comment by Duankaiwen** [comment]
@Mr2er0  cross_iou_loss是文章的所提出的loss， pred：torch.Size([25600, 148])， target: torch.Size([25600, 148])

**Comment by zlj63501** [comment]
> @Zlj-github
> 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742

谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)*2, *2 应该有x,y坐标，所以*2，但是另外一个*2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？

还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~

**Comment by Marcovaldong** [comment]
> > @Zlj-github
> > 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742
> 
> 谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)_2, _2 应该有x,y坐标，所以_2，但是另外一个_2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？
> 
> 还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
> 有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~

hi 你搞懂为啥是(36+1)*4，而不是(36+1)*2了吗？
https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/models/losses/cross_iou_loss.py#L37  这一行resize完之后是K x 72 x 2，再看下面两行，感觉是分成了正负样本两类？0所对应的是负样本，是这样吗？

**Comment by Duankaiwen** [comment]
> > @Zlj-github
> > 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742
> 
> 谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)_2, _2 应该有x,y坐标，所以_2，但是另外一个_2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？
> 
> 还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
> 有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~

@Zlj-github 
每个向量预测四个分量，因此是4倍的关系，详情请看 https://zhuanlan.zhihu.com/p/366651996
这一版的loss形式参考了 [CIOU loss](https://arxiv.org/abs/2005.03572)， 实际上，loss完全可以写成loss =  1 – overlaps，差别不大

**Comment by Duankaiwen** [comment]
> > > @Zlj-github
> > > 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742
> > 
> > 
> > 谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)_2, _2 应该有x,y坐标，所以_2，但是另外一个_2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？
> > 还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
> > 有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~
> 
> hi 你搞懂为啥是(36+1)*4，而不是(36+1)*2了吗？
> https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/models/losses/cross_iou_loss.py#L37 这一行resize完之后是K x 72 x 2，再看下面两行，感觉是分成了正负样本两类？0所对应的是负样本，是这样吗？

@Marcovaldong 
因为每个向量预测四个分量，因此是4倍的关系，详情请看 https://zhuanlan.zhihu.com/p/366651996

**Comment by zlj63501** [comment]
感谢作者大大~

**Comment by Marcovaldong** [comment]
> > > > @Zlj-github
> > > > 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742
> > > 
> > > 
> > > 谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)_2, _2 应该有x,y坐标，所以_2，但是另外一个_2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？
> > > 还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
> > > 有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~
> > 
> > 
> > hi 你搞懂为啥是(36+1)*4，而不是(36+1)*2了吗？
> > https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/models/losses/cross_iou_loss.py#L37 这一行resize完之后是K x 72 x 2，再看下面两行，感觉是分成了正负样本两类？0所对应的是负样本，是这样吗？
> 
> @Marcovaldong
> 因为每个向量预测四个分量，因此是4倍的关系，详情请看 https://zhuanlan.zhihu.com/p/366651996

谢谢老师的回复。我在尝试把cross iou loss加到CenterNet上来，但没有搞懂这个gt的构造。
我现在可以知道https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/datasets/pipelines/loading.py#L422可以得到36个坐标点的标注，而L515这儿是返回了一张图的所有信息。但是我没有找到将36个坐标点转换成36x4个坐标分量的代码在哪儿？您可以告知一下吗？谢谢。

您的这篇论文思路非常棒，感觉无论是在学术界还是在工业上，都还有很大的空间。

**Comment by Marcovaldong** [comment]
> > > > > @Zlj-github
> > > > > 36+1其中36是指每个anchor点预测36根向量，1是指每个anchor点还额外预测了1个向量，这个向量指向物体的几何中心点，该向量可以作为DCN的其中一个offset，因此可以把几何中心点附近的特征给提取到（文章只说了利用DCN提取极值点附近的特征，实际这一版代码中还同时提取了几何中心点附近特征）。至于target (2(batch)hw, 742)中的742在哪个地方?应该也是742
> > > > 
> > > > 
> > > > 谢谢回复，742是打字没有注意符号 74 * 2, 74 =(36+1)_2, _2 应该有x,y坐标，所以_2，但是另外一个_2怎么理解呢？是论文中取0.2倍后的4维坐标系嘛？
> > > > 还有源码里面这个地方： loss = 1 - (overlaps - (rho2 / c2 + v**2 / (1 - overlaps +v)))
> > > > 有点迷惑，好像对不上论文上的公式(出现了arctan函数), 我给您的outlook邮箱发了邮件，您查看一下，谢谢您~
> > > 
> > > 
> > > hi 你搞懂为啥是(36+1)*4，而不是(36+1)*2了吗？
> > > https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/models/losses/cross_iou_loss.py#L37 这一行resize完之后是K x 72 x 2，再看下面两行，感觉是分成了正负样本两类？0所对应的是负样本，是这样吗？
> > 
> > 
> > @Marcovaldong
> > 因为每个向量预测四个分量，因此是4倍的关系，详情请看 https://zhuanlan.zhihu.com/p/366651996
> 
> 谢谢老师的回复。我在尝试把cross iou loss加到CenterNet上来，但没有搞懂这个gt的构造。
> 我现在可以知道[https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/datasets/pipelines/loading.py#L422可以得到36个坐标点的标注，而L515这儿是返回了一张图的所有信息。但是我没有找到将36个坐标点转换成36x4个坐标分量的代码在哪儿？您可以告知一下吗？谢谢。](https://github.com/Duankaiwen/LSNet/blob/main/code/mmdet/datasets/pipelines/loading.py#L422%E5%8F%AF%E4%BB%A5%E5%BE%97%E5%88%B036%E4%B8%AA%E5%9D%90%E6%A0%87%E7%82%B9%E7%9A%84%E6%A0%87%E6%B3%A8%EF%BC%8C%E8%80%8CL515%E8%BF%99%E5%84%BF%E6%98%AF%E8%BF%94%E5%9B%9E%E4%BA%86%E4%B8%80%E5%BC%A0%E5%9B%BE%E7%9A%84%E6%89%80%E6%9C%89%E4%BF%A1%E6%81%AF%E3%80%82%E4%BD%86%E6%98%AF%E6%88%91%E6%B2%A1%E6%9C%89%E6%89%BE%E5%88%B0%E5%B0%8636%E4%B8%AA%E5%9D%90%E6%A0%87%E7%82%B9%E8%BD%AC%E6%8D%A2%E6%88%9036x4%E4%B8%AA%E5%9D%90%E6%A0%87%E5%88%86%E9%87%8F%E7%9A%84%E4%BB%A3%E7%A0%81%E5%9C%A8%E5%93%AA%E5%84%BF%EF%BC%9F%E6%82%A8%E5%8F%AF%E4%BB%A5%E5%91%8A%E7%9F%A5%E4%B8%80%E4%B8%8B%E5%90%97%EF%BC%9F%E8%B0%A2%E8%B0%A2%E3%80%82)
> 
> 您的这篇论文思路非常棒，感觉无论是在学术界还是在工业上，都还有很大的空间。


我想我找到了，https://github.com/Duankaiwen/LSNet/blob/7ca365cefdbab3a039fff160bc61acd0d16bb70c/code/mmdet/models/dense_heads/lsnet_head.py#L1113
但是没看懂这个anchor_pts_list为啥会是一个三维的，而且看起来是平铺到整张图上的？这个意思是不是说标注的mask所包含的36个点都要和对应的proposal的坐标计算出一个向量，这个向量就是loss中要用到的向量。最后再通过normalize_term进行归一化？
